oshi-core/src/test/java/oshi/util/MemoizerTest.java

```diff
@@ -40,7 +39,7 @@ public void after() throws InterruptedException {
         ex.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
     }
 
-    @Test
+    // @Test
```

Comment:
Does this mean the fix proposed here https://github.com/oshi/oshi/pull/959#discussion_r317253433 not helped?

Reply:
Haven't gotten to that.  I'm hunting down the other failing tests involving the Linux C library requiring Unix functions.

Reply:
And... got to it, but it still fails.

Reply:
I still think it's not hitting that last guaranteed iteration when it's interrupted, and is slow when running alongside other tests, so the interruption is more likely to prevent that.  Can we set the interrupted flag and read it later to know whether to decrement the minimum by one?

Reply:
> I still think it's not hitting that last guaranteed iteration when it's interrupted

With the proposed fix, the interrupted flag does not affect the `while` condition. Which means a thread definitely enters the loop at least once, and then if it's interrupted, it throws an `InterruptedException` thus skipping the `m.get()` call. This would have caused observing only a single iteration when asserting at the end of the test, but the thrown exception is re-thrown by `result.get()` which leads to test termination.

Could you tell me how do you run the build to reproduce the failure? I'd like to observe what you see and debug it, but `mvn clean package` finishes successfully for me.

Reply:
The test is only failing on one machine, and only in one situation... using Eclipse, I highlight my project `oshi-core` and select the `src\test\java` package, then "Run as... Junit Test".  It executes all the tests in the project.  It doesn't fail every time, but does at least half the time I try. Might be a hardware limitation (2 physical/4 logical cores).

 Earlier when the tests were failing on Travis it was only on JDK 1.8 (Mac) so that's another possibility.

![Capture](https://user-images.githubusercontent.com/9291703/63629342-77771480-c5c6-11e9-8d0f-f5dfd0114516.JPG)

Reply:
Interestingly when it succeeds it takes longer than when it fails... successes are over 60 seconds, while failures are ~40 seconds.

Reply:
Inconsistent failures like that, often only reproducible on certain systems/environments, are typical for concurrency issues. It often takes specific circumstances for the nanosecond timing to be in such a way that the bug is triggered.

Reply:
@Nadahar
In this case I am not sure this is related to concurrency issues. Flaky behaviour of this test may (and likely is) be caused by unfortunate timings within a single thread.

@dbwiddis 
Are you sure you are running the test after fixing it? The version in the branch `dbwiddis:thread-safety` (commit d66d7ebbbbfaa6be541218f0c7b4412472459d6e) does not contain the fix? May I ask you to commit the version of the test which you are running?

Reply:
Oh, you introduced 98a03a74db08b719acdbe798e905cae327bc0d0e, and I missed it, sorry. So the current version is:
```
while ((System.nanoTime() - beginNanos < iterationDurationNanos
    && !Thread.currentThread().isInterrupted())
    || (guaranteedLastIteration = !guaranteedLastIteration)) {
  if (Thread.currentThread().isInterrupted()) {
    throw new InterruptedException();
  }
...
```
The check of the interrupted flag is still there, and is causing the problem. Could you try with the version I suggested originally?
```
while (System.nanoTime() - beginNanos < iterationDurationNanos
    || (guaranteedLastIteration = !guaranteedLastIteration)) {
  if (Thread.currentThread().isInterrupted()) {
    throw new InterruptedException();
  }
  ...
}
```

Reply:
Ah, sorry, I missed the removal of the middle clause in the surrounding loops and just added the new conditional.  I was too busy debugging silly copy/paste errors instead of actually learning what was happening.   Will make that change and test...

Reply:
Still the same error on my Win10-64 laptop in Eclipse (4 out of 5 times).  Confirmed latest commit:

> HEAD is now at ff51f0187d Fix MemoizerTest

The loop is:
```
while (System.nanoTime() - beginNanos < iterationDurationNanos
        || (guaranteedLastIteration = !guaranteedLastIteration)) {
    if (Thread.currentThread().isInterrupted()) {
        throw new InterruptedException();
    }
...
}
```

Reply:
Latest code also failed the Mac Java 8 build on Travis:
https://travis-ci.org/oshi/oshi/jobs/576107414

Reply:
Uh, I am absolutely failing to see what may be causing this. We know, that either one of the following or both are incorrect:
- `MemoizerTest`;
- `Memoizer.memoize`.

I am suggesting to try playing around with them trying to exclude (to some extend) one or the other out of the equation:
1. set `numberOfThreads = 1` in `MemoizerTest`. Check if the problem reproduces;
2. _undo the changes_, mark `synchronized` the whole `get` method in the `Supplier` produced by `Memoizer.memoize`. Check if the problem reproduces.

If any of the aforementioned changes the reproducibility of the problem, we will at least discover the most plausible culprit and will continue thinking from there. But this is just what I think, maybe you have other ideas.

P.S. I tried reproducing it on JDK 1.8.0_162 (HotSpot), and it still never fails for me, which is not helping.

Reply:
Tried:
- `numberOfThreads = 1` ... problem remains
- synchronizing entire `get()` in the Memoizer... problem remains
- both of the above simultaneously ... problem remains

Other observations:
- I can reproduce the problem at will by causing disk activity from within Eclipse, such as editing and saving a file.  This is interesting because it introduces some latency.
- The comment on the `while()` loop indicates that it is guaranteed to run once -- either because we are under the ttl or because (when that fails) of the guaranteed iteration. (But why not a do...while instead of the boolean? But I digress.) . But what of the situation where we do not reach the `while()` loop until ttl has expired?  We will only get one iteration.

Reply:
To clarify:  the loop will run once by guarantee but is not guaranteed to run twice which the (failing) test assumes.  I have an idea on how to test this... will report back shortly.

Reply:
So there are multiple times the loop only executes once.   These do not all cause failures. But every time it does fail, it is immediately after a one-loop case.

Reply:
> But what of the situation where we do not reach the while() loop until ttl has expired? We will only get one iteration.

Exactly (great catch)! If you introduce a pause (`Thread.sleep`) longer than `iterationDurationNanos` right at the beginning of the task submitted to an executor (i.e. right before `Long previousValue = m.get();`), the problem will be reproduced almost always. Unfortunately, I have some things to do right now, but I will get back to this as soon as I can.

P.S.
> why not a do...while instead of the boolean?
Yes, it very well may be possible. I so rarely use do-while (never), that I forgot it exists...

Reply:
> So there are multiple times the loop only executes once. These do not all cause failures.

These situations don't cause failure when the memoizer is supposed to remember the calculated value forever (ttl is -1). And again, we are counting not the loop iterations, but the number of the calls to the supplier wrapped into a supplier that implements memoization.

Reply:
I think I've figured it out.  You are assuming that the `value` will reach 2:
- Initial call to `m.get()` sets it to 1 before reaching the loop
- In the loop, the second call will increment after ttl has expired.

However, in the case where (a) the loop only executes once, and (b) ttl has not expired, the second call to `m.get()` does not increment the value.

Armed with this information I can cause the failure every time on any system by calling `run()` with `ttlNanos = 1`, forcing the one-iteration case, and relying on catching an instance where code executes faster than `System.nanoTime()` can increment, and thus the `get()` returns the same value.

Reply:
Er... ignore that `ttlNanos = 1`, mental math mistake.  It's the other direction, a long `ttlNanos` equal to the first argument, that reliably fails.

Reply:
I thought you were talking about this same idea initially) And yes, this is exactly the case. And the approach of reproducibility I proposed here (https://github.com/oshi/oshi/pull/959#discussion_r317366400) effectively leads to the same situation you have with `ttl = iterationDurationNanos`.

However, we still want the test to call the original supplier at least twice (this increases our chances of spotting a bug). This means that the condition of the `while` loop must take into account not only the time passed since the iteration started, but also the time passed since we called the original supplier before entering the loop. Something like this:
```
Long previousValue = m.get();
final long firstSupplierCallNanos = System.nanoTime();
assertNotNull(previousValue);
assertTrue(previousValue > 0);
boolean guaranteedLastIteration = false;
// this loop is guaranteed to be executed at least once regardless of the values
// returned by System.nanoTime
while (System.nanoTime() - beginNanos < iterationDurationNanos
    || System.nanoTime() - firstSupplierCallNanos < ttlNanos
    || (guaranteedLastIteration = !guaranteedLastIteration)) {
...
```
And if this works, we can further think about rewriting it with do-while, but it's important to not introduce any occasional additional iterations while doing so, because it in that case w may also need to re-estimate the exact Max value we are asserting.

Reply:
Hmmm remember I altered `s.value` to `s.get() - 1`.   But `s` is the non-threadsafe supplier.  Should I be testing `m.get() - 1` after an enforced delay of `ttlNanos`?

Reply:
The latest suggestion fixed the problem.   I also changed it to a do-while and removed the boolean.   Can't get this to fail....

```
                Long previousValue = m.get();
                final long firstSupplierCallNanos = System.nanoTime();
                assertNotNull(previousValue);
                assertTrue(previousValue > 0);
                // this loop is guaranteed to be executed at least once regardless of the values
                // returned by System.nanoTime
                long now;
                do {
                    if (Thread.currentThread().isInterrupted()) {
                        throw new InterruptedException();
                    }
                    final Long newValue = m.get();
                    assertNotNull(newValue);// check that we never get uninitialized value
                    assertTrue(String.format("newValue=%s, previousValue=%s", newValue, previousValue),
                            newValue >= previousValue);// check that the counter never goes down
                    previousValue = newValue;
                } while ((now = System.nanoTime()) - beginNanos < iterationDurationNanos
                        || now - firstSupplierCallNanos < ttlNanos);

```

Reply:
As I mentioned previously, `s.get() - 1` is safe because the loop doing `Future.get()` provides both a happens-before relation which gives us visibility and guarantees that all other threads that had access to the unsafe supplier are now dormant. So the assertion part of the test has exclusive access to the unsafe supplier.

Doing the same via `m` is less trivia because you would have to figure if it is going to call the underlying supplier or not.

Reply:
> The latest suggestion fixed the problem. I also changed it to a do-while and removed the boolean. Can't get this to fail....

@dbwiddis Great! And looks good also, thank you! I'm so glad this mystery is solved.